File: q23_single_lift_system/solution-1-java/LiftSystem.java
import java.util.*;

/**
 * LiftSystem
 *
 * Single-lift feasibility planner that decides whether to ACCEPT or REJECT a pickup request.
 * - UP and DOWN requests are evaluated on independent passes (they do NOT clash).
 * - For each direction, we plan to stop ONLY at the union of all pickup and drop floors (minimal halts).
 * - K-stop rule (per rider): number of halts strictly between src and dst must be <= maxStops.
 *   (Exclude boarding and destination floors from the count.)
 * - Capacity rule: at any segment between consecutive floors, onboard <= liftCapacity.
 *
 * Data structures per direction:
 * - TreeSet<Integer> stops: unique planned halts for that direction (pickups/drops only).
 * - ArrayList<Rider> riders: accepted riders in that direction (intervals [min(src,dst), max(src,dst)]).
 *
 * Acceptance check for a new request (src, dst):
 * 1) Build proposedStops = stops ∪ {src, dst}  (coalesces duplicates).
 * 2) Verify K-stop rule for each existing rider and the new rider using proposedStops.
 *    For rider interval (low, high), counted halts = |proposedStops ∩ (low, high)|.
 * 3) Capacity feasibility via line sweep:
 *    - Build diff array: diff[low] += 1; diff[high] -= 1 for every rider (existing + new).
 *    - Running sum across floors gives active riders on each travel segment; must be <= capacity.
 *
 * If both checks pass → accept (add rider and stops). Otherwise reject.
 */
public class LiftSystem {

    private final int floorsCount;
    private final int capacity;
    private final int maxStops;

    private final Planner up;    // handles requests with source <= destination
    private final Planner down;  // handles requests with source >  destination

    public LiftSystem(int floorsCount, int liftCapacity, int maxStops) {
        if (floorsCount < 2 || floorsCount > 100) {
            throw new IllegalArgumentException("floorsCount out of range");
        }
        if (liftCapacity < 1 || liftCapacity > 20) {
            throw new IllegalArgumentException("liftCapacity out of range");
        }
        if (maxStops < 1 || maxStops > 10) {
            throw new IllegalArgumentException("maxStops out of range");
        }
        this.floorsCount = floorsCount;
        this.capacity = liftCapacity;
        this.maxStops = maxStops;
        this.up = new Planner();
        this.down = new Planner();
    }

    /**
     * Submit a new ride request. Returns true if accepted (feasible), else false.
     * Direction rule:
     * - if source <= destination → UP planner
     * - if source >  destination → DOWN planner
     */
    public boolean requestPickup(int source, int destination) {
        if (!inRange(source) || !inRange(destination)) return false;

        // Trivial case: same floor request (no travel); always feasible; no state change needed.
        if (source == destination) return true;

        Planner planner = (source <= destination) ? up : down;
        if (planner.canAccept(source, destination, floorsCount, capacity, maxStops)) {
            planner.accept(source, destination);
            return true;
        }
        return false;
    }

    // Optional: debugging helpers (not required by the spec)
    public Set<Integer> getUpStops()   { return Collections.unmodifiableSet(up.stops); }
    public Set<Integer> getDownStops() { return Collections.unmodifiableSet(down.stops); }
    public int getUpRiderCount()       { return up.riders.size(); }
    public int getDownRiderCount()     { return down.riders.size(); }

    // ---- internals ----

    private boolean inRange(int floor) {
        return 0 <= floor && floor < floorsCount;
    }

    private static final class Rider {
        final int src;
        final int dst;
        final int low;
        final int high;
        Rider(int s, int d) {
            this.src = s;
            this.dst = d;
            this.low = Math.min(s, d);
            this.high = Math.max(s, d);
        }
    }

    private static final class Planner {
        private final TreeSet<Integer> stops = new TreeSet<>();
        private final ArrayList<Rider> riders = new ArrayList<>();

        boolean canAccept(int src, int dst, int floorsCount, int capacity, int maxStops) {
            // 1) Build proposed stops (minimal plan = pickups/drops only)
            TreeSet<Integer> proposedStops = new TreeSet<>(stops);
            proposedStops.add(src);
            proposedStops.add(dst);

            // 2) K-stop check for existing riders
            for (Rider r : riders) {
                int inside = countInteriorStops(proposedStops, r.low, r.high);
                if (inside > maxStops) return false;
            }
            // 2b) K-stop check for the new rider
            int low = Math.min(src, dst), high = Math.max(src, dst);
            if (countInteriorStops(proposedStops, low, high) > maxStops) return false;

            // 3) Capacity check via line sweep
            // diff[f] accumulates net starts(+1) and ends(-1) at floor f
            int[] diff = new int[floorsCount + 1]; // +1 guard for ease; we won't index beyond floorsCount
            for (Rider r : riders) {
                diff[r.low] += 1;
                diff[r.high] -= 1;
            }
            diff[low] += 1;
            diff[high] -= 1;

            int active = 0;
            // Each iteration 'f' represents the segment (f, f+1)
            for (int f = 0; f < floorsCount - 1; f++) {
                active += diff[f];
                if (active > capacity) return false;
            }

            return true;
        }

        void accept(int src, int dst) {
            riders.add(new Rider(src, dst));
            stops.add(src);
            stops.add(dst);
        }

        private static int countInteriorStops(NavigableSet<Integer> stops, int low, int high) {
            if (low >= high) return 0;
            // strictly between low and high
            return stops.subSet(low, false, high, false).size();
        }
    }

    
}

File: q23_single_lift_system/solution-2-python/LiftSystem.py
# Design an Elevator System - Request Feasibility (Single Lift)
# Python reference implementation that enforces:
# - Per-direction independence (UP and DOWN do not clash)
# - Capacity on every floor-to-floor segment within a direction
# - "maxStops" rule for every rider already accepted AND the new rider:
#   Count distinct halt floors strictly between (source, destination).
#   Halts are all accepted sources/destinations in that direction.
# - Inputs are validated against floorsCount and source != destination.
#
# Assumption from spec: the car remains IDLE during submissions; we only
# build/validate a feasible plan of halts. No timing/simulation is needed.

from typing import List, Tuple

class LiftSystem:
    def __init__(self, floorsCount: int, liftCapacity: int, maxStops: int):
        # basic argument checks
        if not (2 <= floorsCount <= 100):
            raise ValueError("floorsCount must be in [2, 100]")
        if not (1 <= liftCapacity <= 20):
            raise ValueError("liftCapacity must be in [1, 20]")
        if not (1 <= maxStops <= 10):
            raise ValueError("maxStops must be in [1, 10]")

        self.floorsCount = floorsCount
        self.liftCapacity = liftCapacity
        self.maxStops = maxStops

        # Maintain accepted requests per direction.
        # UP rides store (src,dst) with src < dst
        # DOWN rides store (src,dst) with src > dst
        self.up_rides: List[Tuple[int, int]] = []
        self.down_rides: List[Tuple[int, int]] = []

    def requestPickup(self, source: int, destination: int) -> bool:
        # Validate inputs
        if not (0 <= source < self.floorsCount and 0 <= destination < self.floorsCount):
            return False
        if source == destination:
            # Spec: "source and destination will never be same"
            return False

        is_up = source < destination
        rides = self.up_rides if is_up else self.down_rides

        # Normalize tuple direction (src < dst for UP, src > dst for DOWN) for internal consistency
        s, d = (source, destination) if is_up else (source, destination)

        # Try adding and check feasibility
        candidate = rides + [(s, d)]
        if not self._feasible(candidate):
            return False

        # If feasible, persist
        rides.append((s, d))
        return True

    # ---------- Internal helpers ----------

    def _feasible(self, rides: List[Tuple[int, int]]) -> bool:
        """
        Check both constraints for ONE direction:
        - Capacity on every segment
        - maxStops for every rider (including the newest)
        """
        # Build the set of distinct halt floors in THIS direction:
        # every source/destination of every accepted request
        halts = set()
        for s, d in rides:
            halts.add(s)
            halts.add(d)

        # maxStops check for EACH rider
        for s, d in rides:
            lo, hi = (s, d) if s < d else (d, s)  # strict bounds
            # Distinct halts strictly between lo and hi
            stops_between = sum(1 for h in halts if lo < h < hi)
            if stops_between > self.maxStops:
                return False

        # Capacity check per floor-to-floor segment
        # segments are [f, f+1) for integer floors f
        load = [0] * (self.floorsCount - 1)  # load[f] is riders crossing segment f→f+1
        for s, d in rides:
            lo, hi = (s, d) if s < d else (d, s)
            for f in range(lo, hi):
                load[f - 0] += 1
                if load[f - 0] > self.liftCapacity:
                    return False

        return True

